### ISSUE TITLE
Testing ..

### ISSUE THREAD

**Issue by Yaredoh** [DrSleep/tensorflow-deeplab-resnet#85]
Hello Drsleep,

How can we test it after training with new images only (without giving it labels)?  I didn't saw **test.py** in the repo. 

test.txt 

`/JPEGImages/2008_000006.jpg

/JPEGImages/2008_000011.jpg
/JPEGImages/2008_000012.jpg
/JPEGImages/2008_000018.jpg
/JPEGImages/2008_000024.jpg
/JPEGImages/2008_000030.jpg
/JPEGImages/2008_000031.jpg
/JPEGImages/2008_000046.jpg`

Thanks

**Comment by Yaredoh** [comment]
Thanks for the information. I will try to implement it.

**Comment by Yaredoh** [comment]
I test my image and the out put is not good. I trained it for 20000 steps. On test.py the test iteration is not working. Did i missed placed any thing on the code?

The out put

`
2017-05-18 16:12:04.005011: I tensorflow/core/common_runtime/gpu/gpu_device.cc:908] DMA: 0 
2017-05-18 16:12:04.005016: I tensorflow/core/common_runtime/gpu/gpu_device.cc:918] 0:   Y 
2017-05-18 16:12:04.005025: I tensorflow/core/common_runtime/gpu/gpu_device.cc:977] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 1070, pci bus id: 0000:03:00.0)
Restored model parameters from snapshots/model.ckpt-20000
step 0
The output file has been saved to ./output/mask1.png

`

The test.py code is

`
from __future__ import print_function

import argparse
from datetime import datetime
import os
import sys
import time

import tensorflow as tf
import numpy as np
from PIL import Image


from deeplab_resnet import DeepLabResNetModel, ImageReader, prepare_label,decode_labels

IMG_MEAN = np.array((104.00698793,116.66876762,122.67891434), dtype=np.float32)

DATA_DIRECTORY = 'dataset/images/'
DATA_LIST_PATH = 'dataset/filenames.txt'
IGNORE_LABEL = 256
NUM_CLASSES = 6
NUM_STEPS = 4 # Number of images in the validation set.
RESTORE_FROM = 'snapshots/model.ckpt-20000'
SAVE_DIR = './output/'

def get_arguments():
    """Parse all the arguments provided from the CLI.
    
    Returns:
      A list of parsed arguments.
    """
    parser = argparse.ArgumentParser(description="DeepLabLFOV Network")
    parser.add_argument("--data-dir", type=str, default=DATA_DIRECTORY,
                        help="Path to the directory containing the PASCAL VOC dataset.")
    parser.add_argument("--data-list", type=str, default=DATA_LIST_PATH,
                        help="Path to the file listing the images in the dataset.")
    parser.add_argument("--ignore-label", type=int, default=IGNORE_LABEL,
                        help="The index of the label to ignore during the training.")
    parser.add_argument("--num-classes", type=int, default=NUM_CLASSES,
                        help="Number of classes to predict (including background).")
    parser.add_argument("--num-steps", type=int, default=NUM_STEPS,
                        help="Number of images in the validation set.")
    parser.add_argument("--restore-from", type=str, default=RESTORE_FROM,
                        help="Where restore model parameters from.")
    parser.add_argument("--not-restore-last", action="store_true",
                        help="Whether to not restore last (FC) layers.")
    parser.add_argument("--save-dir", type=str, default=SAVE_DIR,
                        help="Where to save predicted mask.")
    return parser.parse_args()

def load(saver, sess, ckpt_path):
    '''Load trained weights.
    
    Args:
      saver: TensorFlow saver object.
      sess: TensorFlow session.
      ckpt_path: path to checkpoint file with parameters.
    ''' 
    saver.restore(sess, ckpt_path)
    print("Restored model parameters from {}".format(ckpt_path))

def main():
    """Create the model and start the evaluation process."""
    args = get_arguments()
    
    # Create queue coordinator.
    coord = tf.train.Coordinator()
    
    # Load reader.
    with tf.name_scope("create_inputs"):
        reader = ImageReader(
            args.data_dir,
            args.data_list,
            None, # No defined input size.
            False, # No random scale.
            False, # No random mirror.
            args.ignore_label,
            IMG_MEAN,
            coord)
        image = reader.image
    
    image_batch = tf.expand_dims(image, dim=0) # Add one batch dimension.

    # Create network.
    net = DeepLabResNetModel({'data': image_batch}, is_training=False, num_classes=args.num_classes)

    # Which variables to load.
    restore_var = tf.global_variables()
    
    # Predictions.
    raw_output = net.layers['fc1_voc12']
    raw_output = tf.image.resize_bilinear(raw_output, tf.shape(image_batch)[1:3,])
    raw_output = tf.argmax(raw_output, dimension=3)
    pred = tf.expand_dims(raw_output, dim=3) # Create 4-d tensor.
    
            
    # Set up tf session and initialize variables. 
    config = tf.ConfigProto()
    config.gpu_options.allow_growth = True
    sess = tf.Session(config=config)
    init = tf.global_variables_initializer()
    
    sess.run(init)
    sess.run(tf.local_variables_initializer())
    
    # Load weights.
    loader = tf.train.Saver(var_list=restore_var)
    if args.restore_from is not None:
        load(loader, sess, args.restore_from)
    
    # Start queue threads.
    threads = tf.train.start_queue_runners(coord=coord, sess=sess)
    for step in range(args.num_steps):
        preds = sess.run([pred])
        if step % 100 == 0:
            print('step {:d}'.format(step))

    
    msk = decode_labels(preds, num_classes=args.num_classes)
    im = Image.fromarray(msk[0])
    if not os.path.exists(args.save_dir):
        os.makedirs(args.save_dir)
    im.save(args.save_dir + 'mask1.png')
    
    print('The output file has been saved to {}'.format(args.save_dir + 'mask1.png'))
	
    coord.request_stop()
    coord.join(threads)
    
if __name__ == '__main__':
    main()

`

**Comment by Kenneth-X** [comment]
@Yaredoh hi, i am also stuck in testing， i read carefully about your code and where does the "decode_labels" come from? Did you implement it by youself or from some kind of API?
if it's convenient, could you share your testing code and relevant subfunction? 
thank you very much ( ^_^)

**Comment by daobinhuang** [comment]
from deeplab_resnet import DeepLabResNetModel, ImageReader, decode_labels, prepare_label

**Comment by mokii** [comment]
I run the command line 'python evaluate.py' to test the model with the VOC 2012 val set(1447 images), but the mIoU is bad, 0.047, what's wrong with my test?

**Comment by DrSleep** [comment]
@Yaredoh , looks like you are training on your own set. Did you monitor the progress during the training? 
It might be the case that you would need to adapt the hyperparameters (especially learning_rate and batch_size) as they were chosen for the Pascal VOC dataset).
